src/utils/serial.cpp

```diff
@@ -1,20 +1,213 @@
+    {
+        int result;
+        if (!timeout_only)
+        {
+            result = ctx.set_line_property (BITS_8, STOP_BIT_1, NONE);
+            if (result != 0)
+            {
+                log_error ("set_serial_port_settings");
+                return (int)SerialExitCodes::SET_PORT_STATE_ERROR;
+            }
+            result = set_custom_baudrate (115200);
+            if (result != (int)SerialExitCodes::OK)
+            {
+                return result;
+            }
+            result = ctx.set_modem_control (Ftdi::Context::Dtr | Ftdi::Context::Rts);
+            result |= ctx.set_flow_control (SIO_DISABLE_FLOW_CTRL);
+            if (result != 0)
+            {
+                // -1 setting failed, -2 usb device unavailable
+                log_error ("set_serial_port_settings");
+                return (int)SerialExitCodes::SET_PORT_STATE_ERROR;
+            }
+        }
+
+        ctx.set_usb_read_timeout (ms_timeout);
+
+        return (int)SerialExitCodes::OK;
+    }
+
+    int set_custom_baudrate (int baudrate)
+    {
+        switch (ctx.set_baud_rate (115200))
+        {
+            case 0:
+                return (int)SerialExitCodes::OK;
+            case -3: // usb device unavailable
+                log_error ("set_custom_baudrate");
+                return (int)SerialExitCodes::OPEN_PORT_ERROR;
+            default: // -1 invalid baudrate, -2 setting baudrate failed
+                log_error ("set_custom_baudrate");
+                return (int)SerialExitCodes::SET_PORT_STATE_ERROR;
+        }
+    }
+
+    int flush_buffer ()
+    {
+        switch (ctx.tcflush (Ftdi::Context::Input | Ftdi::Context::Output))
+        {
+            case 0:
+                return (int)SerialExitCodes::OK;
+            case -3: // usb device unavailable
+                log_error ("flush_buffer ()");
+                return (int)SerialExitCodes::OPEN_PORT_ERROR;
+            default: // -1,-2 chip failed to purge a buffer
+                log_error ("flush_buffer ()");
+                return (int)SerialExitCodes::SET_PORT_STATE_ERROR;
+        }
+    }
+
+    int read_from_serial_port (void *bytes_to_read, int size)
+    {
+        // it looks like the ftdi will send us data after its latency (max 255ms) times out,
+        // even if its buffer is empty, so the read is repeated until the timeout is reached.
+
+        auto deadline = std::chrono::steady_clock::now () +
+            std::chrono::milliseconds (ctx.get_usb_read_timeout ());
+        unsigned char *next_bytes = static_cast<unsigned char *> (bytes_to_read);
+        do
+        {
+            int bytes_read = ctx.read (next_bytes, size);
+            // TODO: negative values are libusb error codes, -666 means usb device unavailable
+            if (bytes_read < 0)
+            {
+                log_error ("read_from_serial_port");
+                return bytes_read;
+            }
+            next_bytes += bytes_read;
```

Comment:
And I dont fully understand this statement in while loop `next_bytes == bytes_to_read`

Reply:
thanks for looking at these things !

Unfortunately it respects the timeout only for the usb bus.  The ftdi chip will actually send an empty data packet after its configured "latency" delay expires, which the library treats as a response.  See http://www.ftdichip.com/Support/Documents/AppNotes/AN232B-04_DataLatencyFlow.pdf page 8 item 4; there's also more information elsewhere in htat document.  Even if the usb timeout is set to 1 second, the chip will usually reply with its empty data buffer after only 16 ms.

`next_bytes == bytes_to_read` is true when no data has been received yet.  I can probably add a flag to make that clearer.

Reply:
"next_bytes == bytes_to_read is true when no data has been received yet. I can probably add a flag to make that clearer"
yep, I realise it. But should there be a check that size > 0 in this loop?

Reply:
I pushed another commit where the loop is recoded to clarify that it stops as soon as size > 0 .  I think I hadn't made up my mind whether to do that, when I first coded it.

Reply:
oh! yes there is an edge case if 0 is passed to the function.  i'll try to address that in a second push.

Reply:
I mean that in a loop you read bytes and decrease size variable by bytes read so loop should be stopped when its 0

Reply:
at that point `next_bytes == bytes_to_read` would have been false, stopping the loop.  is the new loop structure clearer?  it's intended to only performs 1 read that returns data.  it could loop until size is read, doesn't seem to really matter.